pkg/timelock/timelock.go

```diff
@@ -128,166 +128,256 @@ func NewTimelockWorker(nodeURL, timelockAddress, callProxyAddress, privateKey st
 // Listen is the main function of a Timelock Worker, it subscribes to events using the ethClient
 // and targeting the contract address set.
 func (tw *Worker) Listen(ctx context.Context) error {
-	if err := ctx.Err(); err != nil {
+	c, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Log timelock-worker configuration.
+	tw.startLog()
+
+	// Handle OS signals to properly stop/kill the process.
+	stopCh := make(chan string)
+	defer close(stopCh)
+	go handleOSSignal(stopCh)
+
+	// Run the scheduler to add/del operations in a thread-safe way.
+	go tw.runScheduler(c)
+
+	// Retrieve historical logs.
+	history, err := tw.retrieveHistoricalLogs(c)
+	if err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
 		return err
 	}
 
-	stopCh := make(chan string)
+	// Create a subscription and retrieve new logs asynchronously.
 	logCh := make(chan types.Log)
-	sigCh := make(chan os.Signal, 1)
-	signal.Notify(sigCh, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	defer close(logCh)
 
-	tw.updateSchedulerDelay(time.Duration(tw.pollPeriod) * time.Second)
-	go tw.runScheduler(ctx)
-	tw.startLog()
+	if err := tw.subscribeNewLogs(c, logCh); err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
+		return err
+	}
 
-	// Shutdown gracefully based on OS interrupts.
-	go func() {
-		for {
-			handleOSSignal(<-sigCh, stopCh)
-		}
-	}()
+	// Main goroutine; processes old and new logs and handles cancellation.
+	tw.processLogs(c, history, logCh, stopCh, cancel)
+
+	// Block until all goroutines are done.
+	wg.Wait()
```

Comment:
The original `wg` is defined at pkg level, so it's shared across every func/method. I'm also thinking about including it in the worker, but it's not really idiomatic to pass around a wg. Since we use Worker as a pointer it's probably good tho.